### ISSUE TITLE
Empty Squares in generated SVGs

### ISSUE THREAD

**Issue by ronyeh** [vexflow/vexflow#71]
`grunt generate:current` doesn't load the fonts properly for the `*.svg.png` output (I get empty squares □□□). 

I think it might be because the `generate_images_pptr.js` script makes an SVG, and then asks puppeteer to convert that SVG to a PNG. However, SVGs cannot load fonts from `node_modules`, so if the user does not have the correct fonts on their OS, they will see empty squares..... :-(

I also get empty squares □□□ in the `*.svg` files, but this makes sense, since the music fonts are NOT installed locally on my machine. (I am testing a real world situation where the user has zero music fonts on their computer.) IMO this is the biggest weakness of V5 at the moment. It does not make portable SVG files since it uses the glyph code points directly, and not the paths.

_Originally posted by @ronyeh in https://github.com/vexflow/vexflow/issues/61#issuecomment-1692972424_

**Comment by rvilarl** [comment]
@ronyeh  we could embed them  https://dzx.fr/blog/svg-font-embedding/

**Comment by ronyeh** [comment]
Yes, you can embed full or partial fonts into the svg as base64, but this needs to be optional, otherwise every SVG generated by vexflow will include its own copy of Bravura.

**Comment by rvilarl** [comment]
@ronyeh can this be closed?

**Comment by ronyeh** [comment]
No this cannot be closed.

This issue is regarding the SVG image that is generated. V4 included the paths, so you could download the SVG and send it to someone else and it would still look perfect. V5 does not include the paths, so if you download the SVG from the webpage and send it to someone, 99% of the time they will see blank squares.

Only people who have the font installed on their OS will be able to view these SVGs. This is a drawback of the V5 approach.

The only way to fix this issue is to have an optional flag that renders paths to the SVG (by using opentype.js).

**Comment by rvilarl** [comment]
See 

Using fonts with Object tags

In 
 https://vecta.io/blog/how-to-use-fonts-in-svg

**Comment by rvilarl** [comment]
See 

Using fonts with Object tags

In 
 https://vecta.io/blog/how-to-use-fonts-in-svg

Y

**Comment by rvilarl** [comment]
![svg_Accidental Automatic_Accidentals Bravura](https://github.com/vexflow/vexflow/assets/22865285/cf4bea6b-5565-456a-b9b3-1690d92f6d55)

**Comment by rvilarl** [comment]
The boxes are there! It complains about a style-src default-src policy :(

**Comment by jaredjj3** [comment]
That's probably related to Content Security Policy (CSP). If you control the server serving the fonts, then you need to update  CSP to allow the font source. 

For example: Content-Security-Policy: default-src 'self'; style-src 'self' https://example.com/fonts/; font-src 'self' https://example.com/fonts/;

Alternatively, you can inline the font in base64 format and inline it directly into the SVG. This way, the font will not be loaded from an external source, and you'll bypass the CSP restrictions. However, this can increase the SVG file size past intolerably.

**Comment by rvilarl** [comment]
> ![svg_Accidental Automatic_Accidentals Bravura](https://user-images.githubusercontent.com/22865285/278139464-0b383e5f-dcaa-4fc9-97b5-a0dd9fe02e80.svg)

however opening a local copy of the file directly in the browser works and once you load it once it also works without internet connection as the font is in cache. I have tried with Windows with no font install.

**Comment by mscuthbert** [comment]
I definitely think that an optional flag to the Renderer to have it render paths would be really helpful -- it might even be default "true" for v5 and make it default "false" in v6, but in any case it's something that will make migration of existing projects be much, much safer; there are enough other breaking changes that anything which can make the upgrade path simpler will help keep the project developing.

**Comment by rvilarl** [comment]
I believe that this is only an issue for the test generated SVGs. Going back to rendering paths  would be really a pity in regards to size of the lib and lost of support of additional fonts.

**Comment by jsmcx** [comment]
As a user of Vexflow, I cannot consider using v5 until this is fixed. I really don't understand why anyone would think a smaller broken lib is better than a larger one that works?

**Comment by rvilarl** [comment]
@jsmcx the  library is not broken, the library uses the fonts directly. If you want to use a font you either install it or download it on the web site using vexflow.
The issue here is related to development, the svg file generated as part of testing do not contain the font and if you want to open them you need your fonts installed like you would need for any text file.

**Comment by ronyeh** [comment]
@jsmcx Thanks for your feedback! This is super valuable.

Do you have an example page or other implementation that we can look at to see if there's a possible solution with V5? Does your solution require all the paths to be included into each SVG produced by vexflow? What if the font glyphs were embedded into each SVG? (Embedding partial fonts might work as an alternative to including the full path each time a VexFlow glyph is drawn.)

**Comment by jsmcx** [comment]
This repository shows how I can very easily generate an SVG with Vexflow v4.x.
[Vexflow SVG export demo](https://github.com/jsmcx/Vexflow-SVG-export)

My full project is much more complex than this, but I'm showing the minimal viable code here in order to get the point across. As a reminder, the SVG generation is just the starting point. I am not well-versed in the technicalities of SVG creation, and thanks to Vexflow v4.x _I don't have to be._

Vexflow v4.x makes it easy even for a non-expert like me. It's a few lines of code, and I don't need to know anything about embedding fonts, base64 encoding, or whatever. It just works.

The vexflow5_basic file is my best attempt at making this work in v5. It's much more complex, and despite my best efforts, it still doesn't work. The exported file contains boxes where the noteheads should be. I've worked at this for weeks now, and I just can't get it to work. I have given up and will be sticking to v4.

To be clear: _I'm not asking you to solve my SVG embedding problem, I'm pointing out that the v5 library has created this problem that I didn't have before._ V5 has broken my workflow, and therefore from the user's perspective, _which is frankly the only perspective that really matters_, Vexflow 5 is a broken library. 

![vexflow-output-v4](https://github.com/user-attachments/assets/63dce719-6148-4512-8e52-f4abdb2a9d2c)
![vexflow-output-v5](https://github.com/user-attachments/assets/a7631c55-68e7-4322-b7f4-a83283b97261)

**Comment by jsmcx** [comment]
> @jsmcx the library is not broken, the library uses the fonts directly. If you want to use a font you either install it or download it on the web site using vexflow. The issue here is related to development, the svg file generated as part of testing do not contain the font and if you want to open them you need your fonts installed like you would need for any text file.
> 
> See working examples at https://vexflow.github.io/vexflow-examples/

While I understand the reasoning behind using fonts directly, this approach adds complexity for users who are not deeply familiar with font management. The embedded SVG glyphs in Vexflow v4.x provided a seamless and user-friendly experience.

Could you consider reintroducing embedded SVG glyphs or providing a simpler way to ensure consistent rendering without requiring manual font installation? This would greatly enhance usability for non-expert users.

**Comment by mscuthbert** [comment]
I want to reiterate what I posted 11mo ago: using paths instead of assuming the font is loaded should be at least an option and probably the default for v5. Otherwise there isn't a safe upgrade path for current 1-4 users to 5. I make many SVGs with musical symbols all the time and I'm always careful to embed the paths in them because I know I can't count on my users to have a SMuFL font installed.

**Comment by ronyeh** [comment]
@jsmcx thanks for your feedback. This is super valuable. 

Your repository link didn't work for me. 

It's a known limitation of the current v5 approach. We will definitely figure out a solution so that users dependent on SVG don't lose out. Our goal is that v5 should have everything that v4 has. We just have to prioritize at the moment, but we will keep these issues open for as long as they are relevant.

**Comment by jsmcx** [comment]
Sorry, my Github settings default to private repos. I've made it public now. There's not a lot to see, just a very basic starting point that shows how I capture the generated image for use elsewhere.

Glyph embedding was one of the strengths of earlier Vexflow versions and greatly simplified the workflow for generating musical SVGs. Reintroducing it as an option in v5 would significantly enhance usability and adoption.

For developers concerned with file size, perhaps a compile flag could be introduced to exclude the embedding of fonts. This would allow those who prioritize smaller file sizes to maintain their preference, while still providing an option for users who need the embedded paths for consistent rendering and ease of use.

**Comment by ronyeh** [comment]
I'm going to paste the v4 compatible code you wrote below so that it's a part of this conversation. Hopefully that's OK. I didn't see an explicit license declaration on your repo. 

```html

<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>VexFlow SVG Example</title>
</head>
<body>
    <div id="output"></div>
    
    <!-- Load library -->
    <script src="https://cdn.jsdelivr.net/npm/vexflow@4.2.2/build/cjs/vexflow.js"></script>
    
    <script>
        const { Factory, EasyScore, System } = Vex.Flow;
        
        const vf = new Factory({
            renderer: { elementId: 'output', width: 500, height: 200 },
        });
        
        const score = vf.EasyScore();
        const system = vf.System();
        
        system
        .addStave({
            voices: [
                score.voice(score.notes('C#5/q, B4, A4, G#4', { stem: 'up' })),
                score.voice(score.notes('C#4/h, C#4', { stem: 'down' })),
            ],
        })
        .addClef('treble')
        .addTimeSignature('4/4');
        
        vf.draw();
    </script>
    
    
    <!-- Download button & script -->
    
    <button id="download">Download SVG</button>
    <script>
        function downloadSVG() {
            const svgElement = document.querySelector('#output svg');
            const svgString = new XMLSerializer().serializeToString(svgElement);
            const url = URL.createObjectURL(new Blob([svgString], { type: 'image/svg+xml' }));
            
            const a = document.createElement('a');
            a.href = url;
            a.download = 'vexflow-output.svg';
            a.click();

            URL.revokeObjectURL(url);
        }
        
        document.getElementById('download').addEventListener('click', downloadSVG);
    </script>
</body>
</html>

```

**Comment by jsmcx** [comment]
To address the concern from @rvilarl regarding users who want the option to use alternate fonts: these people are by definition advanced users who will also be the same people who will not mind dealing with font embedding, base64 encoding, etc. 

Such technicalities are a headache for those who are just beginning to explore this area. Please don't lose sight of those less technically-adept users.

**Comment by jsmcx** [comment]
Embedding fonts or embedding outlines would both work, but please ensure that the user doesn't have to do this manually. I have to admit that after a lot of reading I still don't understand font embedding, and you can bet your bottom dollar that any musicians I'm collaborating with would have even less knowledge on this front.

**Comment by rvilarl** [comment]
@jsmcx how do you generate the SVGs with VexFlow? I would like to follow the same steps and figure out a solution.

**Comment by rvilarl** [comment]
Hi all,
I understand the problem now. I was not aware of the svg generation use case, it is clearly described in this issue but I was to focus on other topics. I thought that it was just a development use case with impact only on testing. I will use the code above together with the bach_test and produce 5 svgs:
- 4.2.2
- 5.0.0-alpha 4 with bravura font url 
- 5.0.0-alpha 4 with embedded base64 complete bravura font 
- 5.0.0-alpha 4 with embedded base64 bravura font (subset available in 4.2.2)
- 5.0.0-alpha 4 with embedded base64 bravura font (subset used by bach)
With these, we can decide what is the best way to go.

**Comment by ronyeh** [comment]
OK, so every time a glyph needs to be drawn, `SVGContext.fillText(text, x, y)` is called. 

I propose that `fillText()` examines the `text` parameter to see which SMuFL glyphs are included. It adds these glyphs to a set, and makes sure to embed these glyphs (in the currently selected font) into the SVG.

That way, any SVG generated by VexFlow will have a subset of Bravura/Petaluma/etc embedded into it.

I'll try to find some time to look into this PR. I'm hoping this approach is sound. 🤞

BTW, if anyone else wants to also work on this issue in parallel, feel free! We can get multiple solutions and pick the best one (or combine the best parts of each solution).

**Comment by rvilarl** [comment]
@ronyeh @mscuthbert @mscuthbert @jsmcx @AaronDavidNewman I have generated SVGs for the Bach test. Could you confirm if you see it properly or with empty squares.

VexFlow4, not using fonts, 135K
![vexflo4](https://github.com/user-attachments/assets/0dceb5b3-3396-4a54-95d8-42d5bd746854)

VexFlow5, with subset from Bravura, 84K
![vexflow5subset](https://github.com/user-attachments/assets/771edba5-f9e9-4f69-a03d-051b7cae7c94)

VexFlow5, with full Gonville, 108K
![vexflow5gonville](https://github.com/user-attachments/assets/b18ee43f-04e6-44b8-bf72-c92c9bcaed18)

**Comment by mscuthbert** [comment]
Very cool solution @rvilarl that allows for using 99% the same code but just choosing to embed the entire font as a base64 data or not.  :-)

Vexflow 4 works perfect (of course!)
Bravura works perfectly (I do have Bravura installed on my own computer, so that font always works)
Gonville works almost perfectly.  The tempo marking "dotted half" (=66) renders a box for the half not and for the dot.

Great work.  I think that this should work pretty well -- just need a setting to choose to embed the fonts in the SVG or not.

(As long as we're creating such a setting, maybe we could also have one that decides whether to make a white background to the SVG or not -- for us dark mode people...)

**Comment by rvilarl** [comment]
@mscuthbert in the ´Bravura´ example I changed the font manually to ´Bravura1´ to make sure that it was not using an installed one :). The Gonville error is expected, Gonville does not include these Glyphs. V4 was rendering a scaled version of the standard note and dot.

**Comment by jsmcx** [comment]
These look good as inline images. What happens when you click on the images (i.e. view the SVGs fullscreen in the browser)? For me, I see this:

VexFlow4, not using fonts:

![Screenshot 2024-07-25 at 10 49 42](https://github.com/user-attachments/assets/ed6ad6d7-59d0-4ce7-b54b-91ee55150505)

VexFlow5, with subset from Bravura:

![Screenshot 2024-07-25 at 10 49 55](https://github.com/user-attachments/assets/59210499-59d6-486e-94c4-52152fd1f3eb)

VexFlow5, with full Gonville:

![Screenshot 2024-07-25 at 10 50 07](https://github.com/user-attachments/assets/a62d8032-e8f6-4c4d-8cb4-d454f7e6f8c8)

I'm on macOS 13.6.7, if that makes any difference.
Chrome Version 126.0.6478.183 (Official Build) (arm64), but same issue occurs in Safari Version 17.5 (18618.2.12.111.5, 18618)

**Comment by jsmcx** [comment]
> (As long as we're creating such a setting, maybe we could also have one that decides whether to make a white background to the SVG or not -- for us dark mode people...)

Yes, the lack of a white background is a problem that affects all versions of Vexflow. Here's how @rvilarl's post looks in dark mode:

![Screenshot 2024-07-25 at 11 03 16](https://github.com/user-attachments/assets/0469e384-05c1-4b60-a687-4ebbee54eb4d)

**Comment by rvilarl** [comment]
@jsmcx thanks for reporting, same porblem on Linux

![image](https://github.com/user-attachments/assets/043cc586-0861-4456-8b92-27d9aa5beba5)

It is a Content Security Policy issue. @jaredjj3 @ronyeh are you familiar with that?

**Comment by AaronDavidNewman** [comment]
This is interesting - are you extracting only the glyphs that you need for the svg and that is why the file size is so small?

You should be able to embed the font information directly in the SVG, using the defs element.  That would fix the CORS issue since everything would be in one file.

It doesn't render any faster and it doesn't fix the bounding box problem, but the self-contained, embedded SVG would be nice for an Ossia section.

**Comment by rvilarl** [comment]
@AaronDavidNewman thanks for the feedback

> This is interesting - are you extracting only the glyphs that you need for the svg and that is why the file size is so small?

I used a subset (needed glyphs) for Bravura, the example with Gonville is the full font. The full font for Bravura occupies 330K, the subset 6K

> You should be able to embed the font information directly in the SVG, using the defs element. That would fix the CORS issue since everything would be in one file.

The font information is already embedded. It is not a CORS issue it is a CSP issue

> It doesn't render any faster and it doesn't fix the bounding box problem, but the self-contained, embedded SVG would be nice for an Ossia section.

I thought the bounding box problem was resolved.

**Comment by rvilarl** [comment]
> @jsmcx thanks for reporting, same porblem on Linux
> 
> ![image](https://private-user-images.githubusercontent.com/22865285/352165858-043cc586-0861-4456-8b92-27d9aa5beba5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5Njg1NzgsIm5iZiI6MTcyMTk2ODI3OCwicGF0aCI6Ii8yMjg2NTI4NS8zNTIxNjU4NTgtMDQzY2M1ODYtMDg2MS00NDU2LThiOTItMjdkOWFhNWJlYmE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI2VDA0MzExOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThjZmUwNmJiMmRhN2NjMjMxNWZmMmY1MzgzYjM2MWY5YmRiOWZmMGI2YjA1MzlhMWZlZDAzNzg4OWM5ZTAxZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0uGQ3XNdWm-WLJeC_fIuwTOewcZlY0MS2lwq01K0140)
> 
> It is a Content Security Policy issue. @jaredjj3 @ronyeh are you familiar with that?

Seems to be a known github topic https://stackoverflow.com/questions/65433947/why-custom-headers-for-content-security-policy-not-working-on-github

**Comment by rvilarl** [comment]
For subsetting bravura I:
- placed the svg between `<html></html>` labels
- used https://github.com/zachleat/glyphhanger to get the subsetted font
- used `base64` to get the base64 version of the font
- manually embedded the base64 version of the font and removed the `<html></html>` labels

**Comment by jsmcx** [comment]
I think I have a solution. Posting here to see if it works.

The examples posted before used characters that are not visible to me as I don't have music fonts installed (this will be a common situation for users). I used a python script to convert these to their corresponding Unicode values. 

Also, instead of base64 encoding the font and embedding it directly within the SVG file, I opted to embed the font outlines in the <defs> section of the SVG. This approach allowed for more direct control over the font rendering and avoided potential issues with data URI limitations or base64 encoding.

![music_render_unicode_test](https://github.com/user-attachments/assets/12af04a9-1e3b-4404-9704-129b7c68fff9)

**Comment by jsmcx** [comment]
Trying to fix this issue, I've not yet succeeded in getting around the Content Security Issue, but I have made 2 changes that should make it easier to work on this problem:

1. The examples posted above use characters that aren't visible without the font installed (which is common - most normal people don't use music fonts). I used a Python script to convert these non-displaying characters into their corresponding Unicode values. 
2. Instead of base64 encoding the font and embedding it directly within the SVG file, I opted to embed the font outlines in the <defs> section of the SVG. This is easier to understand than the base64 embedding.

Unfortunately display is still mangled by GitHub's CSP. Hopefully this might be a step closer to an answer, though.

![music_render_unicode_test](https://github.com/user-attachments/assets/150f3651-f475-4103-9863-3f021b7a5c05)

**Comment by jsmcx** [comment]
In case anyone needs it, the convert-to-unicode script was as follows:

```python
#!/usr/bin/env python3

import re

file_path = "" # path to source file

with open(file_path, 'r', encoding='utf-8') as file:
	svg_content = file.read()
	
# Function to replace non-printable characters with their Unicode equivalents
def replace_non_printable(text):
	return ''.join(f'&#x{ord(char):04x};' if ord(char) > 127 else char for char in text)

# Find all text elements with non-printable characters
text_elements = re.findall(r'(<text[^>]*>)(.*?)(</text>)', svg_content, re.DOTALL)

# Replace non-printable characters in the text elements
for start_tag, original_text, end_tag in text_elements:
	replaced_text = replace_non_printable(original_text)
	new_element = f'{start_tag}{replaced_text}{end_tag}'
	svg_content = svg_content.replace(f'{start_tag}{original_text}{end_tag}', new_element)
	
# Show updated content
print(svg_content)

```

**Comment by jsmcx** [comment]
Another attempt with inline font family instructions (thus removing the need for a <style> definition in the header):

![music_render_unicode_test2](https://github.com/user-attachments/assets/78f008c5-0baa-4d1a-9fac-e7661d2eb82d)

**Comment by rvilarl** [comment]
@jsmcx Thanks for helping!

**Comment by rvilarl** [comment]
@jsmcx based on your code I have made a script (`pyftembed`) that embeds a font subset into a svg
```python
#!/usr/bin/python3

import re
import sys
import subprocess
import base64

file_path = sys.argv[1]
embedded_text = ""

with open(file_path, 'r', encoding='utf-8') as file:
	svg_content = file.read()
	file.close()
	
# Find all text elements
text_elements = re.findall(r'(<text[^>]*>)(.*?)(</text>)', svg_content, re.DOTALL)

# Make a string including all text
for start_tag, original_text, end_tag in text_elements:
	embedded_text = embedded_text + original_text

# Create subset font
with subprocess.Popen(['pyftsubset', sys.argv[2], '--text="' + embedded_text + '"', '--flavor=' + sys.argv[2].split('.')[1], '--output-file=temp' ], stdout=subprocess.PIPE) as proc:
    with open("temp", "rb") as temp_file:
        base64_font = base64.b64encode(temp_file.read())
        
    embedded_style = '<style>@font-face{font-family: "' + sys.argv[2].split('.')[0];
    embedded_style = embedded_style + '";font-style: normal;font-weight: normal;src: url("data:font/' + sys.argv[2].split('.')[1] + ';base64,';
    embedded_style = embedded_style + base64_font.decode('utf-8');
    embedded_style = embedded_style + '");}</style>'
        
    # Find the svg elements
    svg_element = re.findall(r'(<svg[^>]*>)(.*?)(<)', svg_content, re.DOTALL)

    # Embed the font
    svg_content = svg_content.replace(f'{svg_element[0][0]}{svg_element[0][1]}{svg_element[0][2]}', f'{svg_element[0][0]}{svg_element[0][1]}{embedded_style}{svg_element[0][2]}')

    file = open(file_path, "w")
    file.write(svg_content)
    file.close()
```
@ronyeh @mscuthbert @jsmcx @AaronDavidNewman would that be reasonable? Calling `pyftembed vexflow.svg bravura.woff2` to include the subset of the used glyphs in the font.

**Comment by jsmcx** [comment]
Postprocessing adds a step that was unnecessary in v4, and still requires whoever is generating the CSV to have the fonts installed (as above, I can't guarantee availability of a music font on my machine, and I don't think most normal people have even heard of music fonts).

Again, please don't lose sight of the fact that most users are non-technical. Ease of use, and robust display of the output is much more important that file size or library size.

**Comment by ronyeh** [comment]
Thanks for all of this research!

I agree that the solution should be no worse than v4 in terms of usability for the developer. There should be no surprises and it should magically work (since it magically worked before). 

We will have to embed font paths in the svgs by default. And if defs/symbol/use doesn't work with GitHub that means it won't work with other server configurations. (I will verify this again just to make sure.) Worst case, we will produce SVGs that look exactly like they did in v4, with all included paths.

**Comment by jsmcx** [comment]
I think it would be helpful to generate some test SVG files that employed the symbol/use method. I expect that file size will be just as small as embedding a font subset.

**Comment by ronyeh** [comment]
@jsmcx posted: https://github.com/vexflow/vexflow/issues/224#issuecomment-2272933941_

I would view the need for postprocessing as a significant regression. Generated SVGs should be viewable stand-alone with no additional work by the user. Complexity should be handled by the library.


-----------------

@rvilarl posted: https://github.com/vexflow/vexflow/issues/224#issuecomment-2276692101_

I believe we should have different options for the different use cases:
- a web page with several SVGs should probably use font-face to download the fonts once in the html header
- SVGs could embed the font URL using font-face
- SVGs could embed the complete font in base64
- SVGs could embed the used subset of the font in base64, see #236
- SVGs could embed the paths using def
- SVGs could use paths rather than text

I believe that the last 5 options could be resolved with scripts. These scripts could be instantiated automatically.

**Comment by ronyeh** [comment]
@rvilarl thanks for the breakdown of the options. However I think the **default** behavior of VexFlow 5.x should be to embed paths (ideally using def) into the SVGs that are generated by the library. All the other options are fine if the advanced developer wants to optimize page load speed, or font flexibility, etc. 

But we shouldn't forget about web developers who just want to include the vexflow script in their HTML page, and render some SVGs on the page, and allow for those SVGs to be "right click saved" and 100% portable on any system.

**Comment by mscuthbert** [comment]
>  However I think the default behavior of VexFlow 5.x should be to embed paths.

Agreed -- experience again and again has shown that the road to acceptance and upgrades (in a mature project) is to add a cool new feature, but make it optional (v5) .  Let everyone adjust their tools and tests to make sure that the cool new feature works.  Make the new feature easy to opt-into, but a three way switch: "old, new, default" where default = old.  Later on (v6? v7? v9?) make the cool new feature the default.  Then people who haven't fixed their code can do a quick patch to change their code to use "old" and then work on fixing "new".  Finally, only later maybe remove the old way of doing things.

**Comment by rvilarl** [comment]
I have some questions about the concept:
- are the fonts going to be placed in the library? all of them?
- is opentypjs going to placed in the libray? 
- @jsmcx I understood that github is not accepting defs, is that right? 

@ronyeh are you making a PR?

**Comment by jsmcx** [comment]
> * @jsmcx I understood that github is not accepting defs, is that right?

Github's content security policy refuses style sheets. Defs seem to be ok. Symbols seem to be ok. 

Sites with user-hosted content are starting to use increasingly restrictive security policies (rightly so), so this is an issue that can't be ignored. 

The image below uses a single `symbol` definition for a treble clef, which is then used 3 times in different colours. It works fine in my own testing but please let me know if you see any display problems.

![svg_symbol_test](https://github.com/user-attachments/assets/4edd5d99-96bf-4422-81d6-00ea9298aac4)

**Comment by jsmcx** [comment]
> But we shouldn't forget about web developers who just want to include the vexflow script in their HTML page, and render some SVGs on the page, and allow for those SVGs to be "right click saved" and 100% portable on any system.

Yep. There are a lot of people (like me!) who use the Vexflow library regularly, and the fact that we don't have to understand how it works is a wonderful feature that frees up time to work on other things.

**Comment by jsmcx** [comment]
> I believe we should have different options for the different use cases:
> 
> * a web page with several SVGs should probably use font-face to download the fonts once in the html header
> * SVGs could embed the font URL using font-face
> * SVGs could embed the complete font in base64
> * SVGs could embed the used subset of the font in base64, see [embed svg font #236](https://github.com/vexflow/vexflow/pull/236)
> * SVGs could embed the paths using def
> * SVGs could use paths rather than text

Which of these options would be prevented by a strict content security policy?

**Comment by rvilarl** [comment]
> > I believe we should have different options for the different use cases:
> > 
> > * a web page with several SVGs should probably use font-face to download the fonts once in the html header
> > * SVGs could embed the font URL using font-face
> > * SVGs could embed the complete font in base64
> > * SVGs could embed the used subset of the font in base64, see [embed svg font #236](https://github.com/vexflow/vexflow/pull/236)
> > * SVGs could embed the paths using def
> > * SVGs could use paths rather than text
> 
> Which of these options would be prevented by a strict content security policy?

I believe that our target should not be to be able to publish on github. Github allows some restricted formats to document / show issues. As an example .html which is the first option is not allowed at all. In github, the 5, 6 seem to work in the page and when clicked. 3, 4 are fine in the page but not when clicked.

Which programs, internet websites are we wishing to address? 
@jsmcx what programs, platforms do you use?

**Comment by ronyeh** [comment]
> are you making a PR?

I am not currently working on a PR for this. Unfortunately I don't have much bandwidth at the moment. I can start looking into this after August 20th. 

I think the right solution would be to use opentype.js to extract the paths at runtime and inject the paths into the SVG context. We can either do it the V4 way by having a path for each SMuFL glyph, or we can do it via defs/symbol and use. I'd prefer the defs approach because it seems nicer. 

Then we can add a flag that is optional, to turn on the new behavior which is to use the glyph unicode directly. This advanced flag is for devs who know what they're doing, and the associated risks of turning off path embedding (I.e. empty squares). 

In the interest of moving forward, as I said in the release plan thread, I think we should target a 5.0 release without this fix. We should just fix any small bugs and add polish to readme and examples and then get 5.0 shipped. 

This SVG issue will be first priority for version 5.1. Anyone who needs perfect SVG functionality can stay on version 4.x for now.

**Comment by jsmcx** [comment]
> Which programs, internet websites are we wishing to address? @jsmcx what programs, platforms do you use?

I'm simply hosting the SVGs on Google Drive for colleagues to see. I think that's a very common use case! Again, and I'm sorry to keep banging on about this, most normal people don't host websites, and they don't know how to install web fonts. 

Here is how the "VexFlow5, with full Gonville" SVG appears when viewed on Google Drive:

![Screenshot 2024-08-10 at 08 40 26](https://github.com/user-attachments/assets/8e75a3af-1fbb-48ae-bc56-d11fd42b5a76)

**Comment by rvilarl** [comment]
@jsmcx is it OK to give me access to your drive so that I can try the options there?

**Comment by rvilarl** [comment]
@ronyeh and what about the fonts, also embedded in the library? All the ones in vexflow-fonts?

**Comment by jsmcx** [comment]
A pragmatic approach would be to have a single default font, stored in whatever way is needed to have outlines embedded in the SVG output. 

People who want multiple fonts and embedded fonts etc could then enable such options and supply the additional files and/or perform the additional work needed.

**Comment by rvilarl** [comment]
I created my own google drive. It seems that SVG security issues have resulted in too many restrictions overall.

https://support.google.com/docs/thread/103766233/images-in-svg-format-no-longer-supported?hl=en

**Comment by jsmcx** [comment]
Sure, but v4 files still display ok. Facebook, Twitter etc have similar security policies. It’s not just GitHub that’s affected by this.

**Comment by rvilarl** [comment]
> > * @jsmcx I understood that github is not accepting defs, is that right?
> 
> Github's content security policy refuses style sheets. Defs seem to be ok. Symbols seem to be ok.
> 
> Sites with user-hosted content are starting to use increasingly restrictive security policies (rightly so), so this is an issue that can't be ignored.
> 
> The image below uses a single `symbol` definition for a treble clef, which is then used 3 times in different colours. It works fine in my own testing but please let me know if you see any display problems.
> 
> ![svg_symbol_test](https://private-user-images.githubusercontent.com/25553382/356479855-4edd5d99-96bf-4422-81d6-00ea9298aac4.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0MTE3NzgsIm5iZiI6MTcyMzQxMTQ3OCwicGF0aCI6Ii8yNTU1MzM4Mi8zNTY0Nzk4NTUtNGVkZDVkOTktOTZiZi00NDIyLTgxZDYtMDBlYTkyOThhYWM0LnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODExVDIxMjQzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxNDczNDQwOWNhYTgzNmViNzNmNTdkMTYxYTE0MjE4MzllMzVhYzA5MzIzYzE5MjIyMTY1ODE0ZmJiMjUwZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dhGZf6PfxvF2yjr1XkbaAX-Z0WqYZqNAFgeBNmhvpWs)

@jsmcx I can see the svg above correctly en github but not on google drive. Have you tried the defs approach in google drive?

**Comment by rvilarl** [comment]
The only SVG approach that works on google drive then is the one from vexflow 4.

It is shocking to see that the security topic has resulted in ignoring styles completely.

**Comment by jsmcx** [comment]
It means that the Vexflow 5 philosophy creates SVGs that display incorrectly on some sites that display user-generated content, such as Google Drive and Github. The same issue may affect Sharepoint, Facebook and Twitter. 

It would be worth including this in the documentation, in order to save end-users from having to go through the same journey of discovery.
